zproject/backends.py

```diff
@@ -1364,6 +1373,74 @@ class GitLabAuthBackend(SocialAuthMixin, GitLabOAuth2):
+    REDIS_EXPIRATION_SECONDS = 60*10
+
+    @staticmethod
+    def get_apple_locale(django_language_code: str) -> str:
+        '''
+        Get the suitable apple supported locale with language code
+        for the Sign in / Continue with Apple buttons it provides.
+        '''
+        # The following is a list of locale values supported by Apple to send
+        # as params to URL which renders "Sign in with Apple" and "Continue with Apple"
+        # buttons. Gathered from
+        # https://developer.apple.com/documentation/signinwithapplejs/incorporating_sign_in_with_apple_into_other_platforms .
+        supported_locales = ['ar_SA', 'ca_ES', 'cs_CZ', 'da_DK', 'de_DE',
+                             'el_GR', 'en_US', 'es_ES', 'fi_FI', 'fr_FR',
+                             'hr_HR', 'hu_HU', 'id_ID', 'it_IT', 'iw_IL',
+                             'ja_JP', 'ko_KR', 'ms_MY', 'nl_NL', 'no_NO',
+                             'pl_PL', 'pt_PT', 'ro_RO', 'ru_RU', 'sk_SK',
+                             'sv_SE', 'th_TH', 'tr_TR', 'uk_UA', 'vi_VI',
+                             'zh_CN']
+        for locale in supported_locales:
+            if django_language_code in locale:
+                return locale
+        return 'en_US'
+
+    # we override the following functions to set state as a redis key with
+    # data as values of variables in `SOCIAL_AUTH_FIELDS_STORED_IN_SESSION`
+    def get_or_create_state(self) -> str:
+        '''
+        Store state in redis for further request validation. The state
+        value is passed as state parameter (as specified in OAuth2 spec),
+        but also added to redirect, that way we can still verify the
+        request if the provider doesn't implement the state parameter.
+        Reuse token if any. This is usually stored in session but, because
+        apple sending user data as POST, storing is session doesn't work.
+        '''
+        request_data = self.strategy.request_data().dict()
+        data_to_store = {
+            key: request_data[key] for key in SOCIAL_AUTH_FIELDS_STORED_IN_SESSION if key in request_data
+        }
+        state = put_dict_in_redis(redis_client, 'apple_auth_{token}',
+                                  data_to_store, self.REDIS_EXPIRATION_SECONDS,
+                                  token = self.state_token())
+        state = state.replace('apple_auth_', '')
+        return state
+
+    def validate_state(self) -> Optional[str]:
+        """Validate state value. Raises exception on error, returns state
+        value if valid."""
+        request_state = self.get_request_state()
+        formatted_request_state = "apple_auth_"+request_state
+        redis_data = get_dict_from_redis(redis_client, "apple_auth_{token}",
+                                         formatted_request_state)
+        if not request_state:
+            raise AuthMissingParameter(self, 'state')
+        elif redis_data is None:
+            raise AuthStateForbidden(self)
```

Comment:
why this exception specifically?

Reply:
the original `validate_state` function had that when both states do not match. So, It is similar here if we do not get redis_data as it means we get a wrong state value.

Reply:
Makes sense!

Reply:
Should we be calling `validate_state` from the superclass rather than reimplementing parts of it?

Reply:
Not really, we're re-doing the whole method. The upstream one compares the `state` in the request with the `state` saved in user's session. But Apple sends the user here with a POST request, so we don't have the session.